### ISSUE TITLE
meijerint error: _CoeffExpValueError: expr not of form `a*x**b`

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#21426]
From SO:
https://stackoverflow.com/questions/67373196/weird-results-from-sympy-integrate-and-heaviside
```python
In [2]: expr = (x - 0.5)*Heaviside(x - 0.5)

In [3]: expr
Out[3]: (x - 0.5)⋅θ(x - 0.5)

In [4]: integrate(expr, (x, x))
---------------------------------------------------------------------------
_CoeffExpValueError                       Traceback (most recent call last)
<ipython-input-4-90dd0a8257c2> in <module>
----> 1 integrate(expr, (x, x))

~/current/sympy/sympy/sympy/integrals/integrals.py in integrate(meijerg, conds, risch, heurisch, manual, *args, **kwargs)
   1569 
   1570     if isinstance(integral, Integral):
-> 1571         return integral.doit(**doit_flags)
   1572     else:
   1573         new_args = [a.doit(**doit_flags) if isinstance(a, Integral) else a

~/current/sympy/sympy/sympy/integrals/integrals.py in doit(self, **hints)
    595                     antideriv = None
    596                 else:
--> 597                     antideriv = self._eval_integral(
    598                         function, xab[0], **eval_kwargs)
    599                     if antideriv is None and meijerg is True:

~/current/sympy/sympy/sympy/integrals/integrals.py in _eval_integral(self, f, x, meijerg, risch, manual, heurisch, conds)
   1079                 # rewrite using G functions
   1080                 try:
-> 1081                     h = meijerint_indefinite(g, x)
   1082                 except NotImplementedError:
   1083                     from sympy.integrals.meijerint import _debug

~/current/sympy/sympy/sympy/integrals/meijerint.py in meijerint_indefinite(f, x)
   1659     results = []
   1660     for a in sorted(_find_splitting_points(f, x) | {S.Zero}, key=default_sort_key):
-> 1661         res = _meijerint_indefinite_1(f.subs(x, x + a), x)
   1662         if not res:
   1663             continue

~/current/sympy/sympy/sympy/integrals/meijerint.py in _meijerint_indefinite_1(f, x)
   1684     _debug('Trying to compute the indefinite integral of', f, 'wrt', x)
   1685 
-> 1686     gs = _rewrite1(f, x)
   1687     if gs is None:
   1688         # Note: the code that calls us will do expand() and try again

~/current/sympy/sympy/sympy/integrals/meijerint.py in _rewrite1(f, x, recursive)
   1607     """
   1608     fac, po, g = _split_mul(f, x)
-> 1609     g = _rewrite_single(g, x, recursive)
   1610     if g:
   1611         return fac, po, g[0], g[1]

~/current/sympy/sympy/sympy/core/cache.py in wrapper(*args, **kwargs)
     70         def wrapper(*args, **kwargs):
     71             try:
---> 72                 retval = cfunc(*args, **kwargs)
     73             except TypeError:
     74                 retval = func(*args, **kwargs)

~/current/sympy/sympy/sympy/integrals/meijerint.py in _rewrite_single(f, x, recursive)
   1509                 res = []
   1510                 for fac, g in terms:
-> 1511                     r1 = _get_coeff_exp(unpolarify(fac.subs(subs).subs(z, x),
   1512                                                    exponents_only=True), x)
   1513                     try:

~/current/sympy/sympy/sympy/integrals/meijerint.py in _get_coeff_exp(expr, x)
    339         return c, S.One
    340     else:
--> 341         raise _CoeffExpValueError('expr not of form a*x**b: %s' % expr)
    342 
    343 

_CoeffExpValueError: expr not of form a*x**b: 0.500000000000000
```
If rationals are used then the result is:
```python
In [5]: expr = (x - S.Half)*Heaviside(x - S.Half)

In [6]: integrate(expr, (x, x))
Out[6]: 
⎧            0               for 2⋅│x│ < 1
⎪                                         
⎪╭─╮0, 2 ⎛3, 1       │    ⎞               
⎨│╶┐     ⎜           │ 2⋅x⎟               
⎪╰─╯2, 2 ⎝      1, 0 │    ⎠               
⎪──────────────────────────    otherwise  
⎩            4 
```
The first part of the Piecewise is correct although the condition should be `x < 1/2` so the result is incorrect for `x < -1/2`.

With float or rational a simpler correct result can be found using Piecewise:
```python
In [28]: expr = (x - 0.5)*Heaviside(x - 0.5)

In [29]: expr.rewrite(Piecewise).integrate(x)
Out[29]: 
⎧          0             for x < 0.5
⎪                                   
⎨     2                             
⎪0.5⋅x  - 0.5⋅x + 0.125   otherwise 
⎩     
```
Although the result comes out different if the limits are given as `(x, x)` which shouldn't really make a difference:
```python
In [36]: expr.rewrite(Piecewise).integrate((x, x))
Out[36]: 
⎧           0              for x - 0.5 < 0
⎪                                         
⎪     2                                   
⎨0.5⋅x ⋅θ(0) - 0.5⋅x⋅θ(0)  for x - 0.5 = 0
⎪                                         
⎪          2                              
⎩     0.5⋅x  - 0.5⋅x       for x - 0.5 > 0
```

**Comment by jksuom** [comment]
it seems that `rational=False` should be added here to extract a floating point coefficient: https://github.com/sympy/sympy/blob/05f5021fff2e74f674e62c56ed0fc58bca17e2d2/sympy/integrals/meijerint.py#L330
```
(Pdb) p expr
0.500000000000000
(Pdb) p expr.as_coeff_mul(x)
(1, (0.500000000000000,))
(Pdb) p expr.as_coeff_mul(x, rational=False)
(0.500000000000000, ())
```

**Comment by laolux** [comment]
Just to add my two cents, the meijerg backend for integrate seems to have some serious issues with the Heaviside function.

    In [47]: sym.integrate(sym.Heaviside(x-1), (x,-2,2))
    Out[47]: -meijerg(((2, 1), ()), ((), (1, 0)), 1) + meijerg(((2, 1), ()), ((), (1, 0)), -1) + meijerg(((2, 1), ()), ((), (1, 0)), 2) - meijerg(((2, 1), ()), ((), (1, 0)), -2)

while it should be:

    In [48]: sym.integrate(sym.Heaviside(x-1), (x,-2,2), meijerg=False)
    Out[48]: 1

What's worse, if we convert the meijerg result to float, then we get

    In [49]: sym.nfloat(sym.integrate(sym.Heaviside(x-1), (x,-2,2)))
    Out[49]: 4.0 + 3.09784953068316e-31*I

which is obviously completely wrong.

**Comment by jksuom** [comment]
`meijerint` can easily handle the *definite* integral:
```
In [39]: meijerint_definite(Heaviside(x-1), x, -2, 2)                           
Out[39]: (1, True)
```
However, it often fails on antiderivatives
```
In [40]: meijerint_indefinite(Heaviside(x-1), x)                                
Out[40]: 
⎧           0              for │x│ < 1
⎪                                     
⎨╭─╮0, 2 ⎛2, 1       │  ⎞             
⎪│╶┐     ⎜           │ x⎟   otherwise 
⎩╰─╯2, 2 ⎝      1, 0 │  ⎠             
```
(This is correct for `x > 0` but not for `x < -1`.)

**Comment by jksuom** [comment]
@laolux You could test this patch:
```
--- a/sympy/integrals/integrals.py
+++ b/sympy/integrals/integrals.py
@@ -1082,6 +1082,9 @@ def _eval_integral(self, f, x, meijerg=None, risch=None, manual=None,
                 except NotImplementedError:
                     from sympy.integrals.meijerint import _debug
                     _debug('NotImplementedError from meijerint_definite')
+                if isinstance(h, Piecewise) and len(self.limits[-1]) == 3:
+                    # special methods are needed for definite integrals
+                    h = None
                 if h is not None:
                     parts.append(coeff * h)
                     continue
```

**Comment by laolux** [comment]
@jksuom Thanks, that patch solves my issue, but not the original issue. Note that I am not the one having posted on stackoverflow. I just happened to have a similar issue with Heaviside, which your patch solves.

**Comment by jksuom** [comment]
@laolux Can you send a PR including a unit test for the patch?

> not the original issue

There is another suggested patch https://github.com/sympy/sympy/issues/21426#issuecomment-831749769 for the original issue.

**Comment by laolux** [comment]
@jksuom with current master e30fcee96c3dbec337bc665229e5ef685277d56e your suggested fix for the original issue does not work.

    expr = (x-0.5)*Heaviside(x-0.5)
    integrate(expr, x)
    Piecewise((0, 2.0*Abs(x) < 1), (0.25*meijerg(((3, 1), ()), ((), (1, 0)), 2.0*x), True))

Shall I go ahead and merge request only the part that works?

**Comment by jksuom** [comment]
> your suggested fix for the original issue does not work.

It seems to fix the error coming from floating point numbers. The result is the same as when rationals are used in Out[6] of the opening paragraph, only floats appear instead of rational numbers.

The result is correct except that the Meijer G-function has not been rewritten. That is the actual issue. (You need not worry about it.)

> Shall I go ahead and merge request only the part that works

Yes, you can add the fix together with a unit test (`integrate(Heaviside(x-1), x, -2, 2)`) even if it does not fix the original issue. The rationale for this fix is as follows: Antiderivatives can be computed as Meijer G-functions but they are not necessarily valid on the negative real line. So, if it is seen that the integral is definite, the result of `meijerint_indefinite` is cancelled in order to make `integrate` use `meijerint_definite`.

**Comment by laolux** [comment]
Sorry, I think I have not been explicit enough why I do not like the result with your `rational=False fix. It leads to a piecewise with `0, 2*Abs(x)<1`, whereas I am pretty sure that the integral should be zero for `0` for any `x<1/2`. And just quickly checking `nfloat(meijerg(((3, 1), ()), ((), (1, 0)), 2.0*(x))) = 12.5 + 2.43624800211422e-25*I` for `x=-2`, so that looks wrong to me. This is also mentioned by oscarbenjamin in the original post.
So long story short, your `rational=False` fix solves part of the issue, but does unfortunately not help with the wrong conditions in the piecewise result.

So I will go ahead and only submit a merge request for the problem affecting me.

**Comment by laolux** [comment]
Submitted merge request #21478, but it only solves part of this issue here, so the issue should not be closed if it gets merged. Just as a reminder.

**Comment by jksuom** [comment]
> a piecewise with 0, 2*Abs(x)<1, whereas I am pretty sure that the integral should be zero for 0 for any x<1/2

This is what I referred to in
> Antiderivatives can be computed as Meijer G-functions but they are not necessarily valid on the negative real line.

`rational=True` was never meant to fix this but only to remove the `_CoeffExpValueError`. There are many places in the integration code that fail with floats. That fix will be needed even if it does not solve the issue with Meijer G-functions.

**Comment by laolux** [comment]
Ok, now I understand, thanks for clarifying.
Well, I would be fine to add this, as it clearly has some benefit, but then I would have to write a test case evaluating to a wrong result. Or omit the test case all together. What should I do?

In any case, I will withdraw and resubmit my merge request, as I just figured out that I misused git...

**Comment by laolux** [comment]
OK, the fix needs more work. It fails a test:

    sympy/functions/elementary/tests/test_piecewise.py", line 786, in test_issue_11045
    assert integrate(1/(x*sqrt(x**2 - 1)), (x, 1, 2)) == pi/3

I doubt that I will be able to fix that.

**Comment by jksuom** [comment]
The next step to try is to be more explicit about negative limits. Consider adding

    and self.limits[-1][1].is_negative

**Comment by laolux** [comment]
Thanks for trying to push me in the right direction to solve the problem, but I am afraid I do not understand the underlying problem, so I will not be able to fix the issue. I simply have no clue about sympy's internals and how symbolic integration is done in software (I do use a lot of pen and paper).

Anyways, I tried your newest suggestion. Works fine only for the special case that the first number is negative.
However, this functionality should work in the general case, for example when some other calculation defines the limits or the like.
So, it should also work when the limits are both positive (trivial solution to the integral) or the lower limit is positive and the second is negative. So basically, it should pass the following:

    a, b, x, y = symbols('a b x y')
    assert simplify(integrate(Heaviside(x - y), (x, a, b)) \
    + (a - y)*Heaviside(a - y, Rational(1, 2)) \
    - (b - y)*Heaviside(b - y, Rational(1, 2))) == 0

Of course, on could restrict the check such that the first integrand has to be negative, but I am not sure if this would be a good solution. Users might be surprised to see that things only work if the first limit is negative.

**Comment by jsdodge** [comment]
I've encountered the same error with the following expression. Any suggestions?

```python
In [1]: from sympy import integrate

In [2]: from sympy.abc import x

In [3]: integrate((1 + x ** 2) ** (-3/2), x)
---------------------------------------------------------------------------
_CoeffExpValueError                       Traceback (most recent call last)
<ipython-input-3-db7acc046ea6> in <module>
----> 1 integrate((1 + x ** 2) ** (-3/2), x)

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in integrate(meijerg, conds, risch, heurisch, manual, *args, **kwargs)
   1550 
   1551     if isinstance(integral, Integral):
-> 1552         return integral.doit(**doit_flags)
   1553     else:
   1554         new_args = [a.doit(**doit_flags) if isinstance(a, Integral) else a

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in doit(self, **hints)
    595                     antideriv = None
    596                 else:
--> 597                     antideriv = self._eval_integral(
    598                         function, xab[0], **eval_kwargs)
    599                     if antideriv is None and meijerg is True:

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in _eval_integral(self, f, x, meijerg, risch, manual, heurisch, conds, final)
   1083                 # rewrite using G functions
   1084                 try:
-> 1085                     h = meijerint_indefinite(g, x)
   1086                 except NotImplementedError:
   1087                     from sympy.integrals.meijerint import _debug

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in meijerint_indefinite(f, x)
   1659     results = []
   1660     for a in sorted(_find_splitting_points(f, x) | {S.Zero}, key=default_sort_key):
-> 1661         res = _meijerint_indefinite_1(f.subs(x, x + a), x)
   1662         if not res:
   1663             continue

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _meijerint_indefinite_1(f, x)
   1684     _debug('Trying to compute the indefinite integral of', f, 'wrt', x)
   1685 
-> 1686     gs = _rewrite1(f, x)
   1687     if gs is None:
   1688         # Note: the code that calls us will do expand() and try again

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _rewrite1(f, x, recursive)
   1607     """
   1608     fac, po, g = _split_mul(f, x)
-> 1609     g = _rewrite_single(g, x, recursive)
   1610     if g:
   1611         return fac, po, g[0], g[1]

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/core/cache.py in wrapper(*args, **kwargs)
     70         def wrapper(*args, **kwargs):
     71             try:
---> 72                 retval = cfunc(*args, **kwargs)
     73             except TypeError as e:
     74                 if not e.args or not e.args[0].startswith('unhashable type:'):

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _rewrite_single(f, x, recursive)
   1509                 res = []
   1510                 for fac, g in terms:
-> 1511                     r1 = _get_coeff_exp(unpolarify(fac.subs(subs).subs(z, x),
   1512                                                    exponents_only=True), x)
   1513                     try:

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _get_coeff_exp(expr, x)
    339         return c, S.One
    340     else:
--> 341         raise _CoeffExpValueError('expr not of form a*x**b: %s' % expr)
    342 
    343 

_CoeffExpValueError: expr not of form a*x**b: 1.12837916709551
```

**Comment by jsdodge** [comment]
Strangely, the definite integral yields a solution, but only if I include a dummy variable in the integrand. The prefactor in the solution is 2/sqrt(pi) = 1.128..., and the standard way to solve the integral is through trigonometric substitution.

```python
In [1]: from sympy import integrate

In [2]: from sympy.abc import x, y

In [3]: integrate((x ** 2 + y ** 2) ** (-3/2), (x, -1, 1))
Out[3]: 1.12837916709551*sqrt(pi)/(y**3.0*sqrt(1 + y**(-2)))

In [4]: integrate((x ** 2 + 1) ** (-3/2), (x, -1, 1))
---------------------------------------------------------------------------
_CoeffExpValueError                       Traceback (most recent call last)
<ipython-input-4-6050e09ca739> in <module>
----> 1 integrate((x ** 2 + 1) ** (-3/2), (x, -1, 1))

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in integrate(meijerg, conds, risch, heurisch, manual, *args, **kwargs)
   1550 
   1551     if isinstance(integral, Integral):
-> 1552         return integral.doit(**doit_flags)
   1553     else:
   1554         new_args = [a.doit(**doit_flags) if isinstance(a, Integral) else a

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in doit(self, **hints)
    484         if reps:
    485             undo = {v: k for k, v in reps.items()}
--> 486             did = self.xreplace(reps).doit(**hints)
    487             if type(did) is tuple:  # when separate=True
    488                 did = tuple([i.xreplace(undo) for i in did])

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in doit(self, **hints)
    595                     antideriv = None
    596                 else:
--> 597                     antideriv = self._eval_integral(
    598                         function, xab[0], **eval_kwargs)
    599                     if antideriv is None and meijerg is True:

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/integrals.py in _eval_integral(self, f, x, meijerg, risch, manual, heurisch, conds, final)
   1083                 # rewrite using G functions
   1084                 try:
-> 1085                     h = meijerint_indefinite(g, x)
   1086                 except NotImplementedError:
   1087                     from sympy.integrals.meijerint import _debug

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in meijerint_indefinite(f, x)
   1659     results = []
   1660     for a in sorted(_find_splitting_points(f, x) | {S.Zero}, key=default_sort_key):
-> 1661         res = _meijerint_indefinite_1(f.subs(x, x + a), x)
   1662         if not res:
   1663             continue

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _meijerint_indefinite_1(f, x)
   1684     _debug('Trying to compute the indefinite integral of', f, 'wrt', x)
   1685 
-> 1686     gs = _rewrite1(f, x)
   1687     if gs is None:
   1688         # Note: the code that calls us will do expand() and try again

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _rewrite1(f, x, recursive)
   1607     """
   1608     fac, po, g = _split_mul(f, x)
-> 1609     g = _rewrite_single(g, x, recursive)
   1610     if g:
   1611         return fac, po, g[0], g[1]

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/core/cache.py in wrapper(*args, **kwargs)
     70         def wrapper(*args, **kwargs):
     71             try:
---> 72                 retval = cfunc(*args, **kwargs)
     73             except TypeError as e:
     74                 if not e.args or not e.args[0].startswith('unhashable type:'):

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _rewrite_single(f, x, recursive)
   1509                 res = []
   1510                 for fac, g in terms:
-> 1511                     r1 = _get_coeff_exp(unpolarify(fac.subs(subs).subs(z, x),
   1512                                                    exponents_only=True), x)
   1513                     try:

~/miniconda3/envs/jsd/lib/python3.9/site-packages/sympy/integrals/meijerint.py in _get_coeff_exp(expr, x)
    339         return c, S.One
    340     else:
--> 341         raise _CoeffExpValueError('expr not of form a*x**b: %s' % expr)
    342 
    343 

_CoeffExpValueError: expr not of form a*x**b: 1.12837916709551

```

**Comment by oscarbenjamin** [comment]
> Any suggestions?

As discussed above. The issue is to do with floats so don't use floats (i.e. `3/2`):
```python
In [5]: integrate((x ** 2 + 1) ** (-S(3)/2), (x, -1, 1))
Out[5]: √2
```

**Comment by jsdodge** [comment]
Thanks, that helps. I did look through the previous discussion but it was all a bit much to digest, since this integral was my first attempt to use SymPy.

May I suggest a more helpful error message, for those of us who are new to the package?
